src/store.js

```diff
@@ -427,3 +427,155 @@ export const useWizardStore = defineStore("wizard", {
+    loading_tools_latency: false,
+    loading_pypi_latency: false,
+
+    // Last updated timestamps (ms epoch)
+    idf_last_updated: 0,
+    tools_last_updated: 0,
+    pypi_last_updated: 0,
+
+    // TTL for latency cache (15 minutes)
+    latency_ttl_ms: 15 * 60 * 1000,
+  }),
+  getters: {
+    idfUrls: (state) => state.idf_urls,
+    toolsUrls: (state) => state.tools_urls,
+    pypiUrls: (state) => state.pypi_urls,
+    idfLatencyMap: (state) => state.idf_latency_map,
+    toolsLatencyMap: (state) => state.tools_latency_map,
+    pypiLatencyMap: (state) => state.pypi_latency_map,
+  },
+  actions: {
+    // Backend uses Option<u32> for latency values; Timedout values are represented as None. 
+    // We normalize to 0 for timeout and the value for the latency. If the value is undefined, we return undefined as it means the mirror is not yet measured.
+    normalizeLatencyValue(value) {
+      if (value === undefined) return undefined;
+      if (value == null) return 0;
+      return Number(value);
+    },
+
+    ttlValid(lastUpdated) {
+      if (!lastUpdated) return false;
+      const now = Date.now();
+      return now - lastUpdated < this.latency_ttl_ms;
+    },
+
+    async bootstrapMirrors() {
+      // Fetch quick URL lists + defaults for all types in parallel
+      this.loading_idf_urls = true;
+      this.loading_tools_urls = true;
+      this.loading_pypi_urls = true;
+      try {
+        const pIdf = invoke("get_idf_mirror_urls", {});
+        const pTools = invoke("get_tools_mirror_urls", {});
+        const pPypi = invoke("get_pypi_mirror_urls", {});
+
+        const [idf, tools, pypi] = await Promise.allSettled([pIdf, pTools, pPypi]);
+
+        if (idf.status === "fulfilled") {
+          const res = idf.value || {};
+          this.idf_urls = Array.isArray(res.mirrors) ? res.mirrors : [];
+          this.selected_idf = typeof res.selected === "string" ? res.selected : "";
+        }
+        if (tools.status === "fulfilled") {
+          const res = tools.value || {};
+          this.tools_urls = Array.isArray(res.mirrors) ? res.mirrors : [];
+          this.selected_tools = typeof res.selected === "string" ? res.selected : "";
+        }
+        if (pypi.status === "fulfilled") {
+          const res = pypi.value || {};
+          this.pypi_urls = Array.isArray(res.mirrors) ? res.mirrors : [];
+          this.selected_pypi = typeof res.selected === "string" ? res.selected : "";
+        }
+      } finally {
+        this.loading_idf_urls = false;
+        this.loading_tools_urls = false;
+        this.loading_pypi_urls = false;
+      }
+
+      // Kick off progressive per-type background latency calculations
+      this.computeLatencyInBackground();
+    },
+
+    computeLatencyInBackground() {
+      const now = Date.now();
+      // IDF
+      if (!this.ttlValid(this.idf_last_updated) && !this.loading_idf_latency) {
+        this.loading_idf_latency = true;
+        invoke("get_idf_mirror_list", {})
```

Comment:
If you refactor Rust to return Vec<MirrorEntry>, then:

The Tauri commands (get_idf_mirror_list, etc.) should return an array.